stdlib/LinearAlgebra/src/blas.jl

```diff
@@ -106,27 +106,78 @@ end
 
 openblas_get_config() = strip(unsafe_string(ccall((@blasfunc(openblas_get_config), libblas), Ptr{UInt8}, () )))
 
+function guess_vendor()
+    # like determine_vendor, but guesses blas in some cases
+    # where determine_vendor returns :unknown
+    ret = vendor()
+    if Sys.isapple() && (ret == :unknown )
+        ret = :osxblas
+    end
+    ret
+end
+
+
 """
-    set_num_threads(n)
+    set_num_threads(n::Integer)
+    set_num_threads(::Nothing)
+
+Set the number of threads the BLAS library should use equal to `n::Integer`.
+If `nothing` is passed to this function, julia tries to figure out the optimial number of threads.
+The exact heuristic is an implementation detail.
 
-Set the number of threads the BLAS library should use.
+On exotic variants of `BLAS` this function can fail.
 """
-function set_num_threads(n::Integer)
-    blas = vendor()
-    if blas === :openblas
-        return ccall((:openblas_set_num_threads, libblas), Cvoid, (Int32,), n)
-    elseif blas === :openblas64
-        return ccall((:openblas_set_num_threads64_, libblas), Cvoid, (Int32,), n)
-    elseif blas === :mkl
+function set_num_threads(n::Integer, _blas=guess_vendor())::Nothing
```

Comment:
is there any reason for exposing the `_blas` argument here. 
Why not
```julia
function set_num_threads(n::Integer)::Nothing
         _blas=guess_vendor()
   ....
```

Reply:
Yeah, its for testing. We can simulate `:osxblas` this way, which is otherwise not covered by CI. See the above discussion with @tkf.

Reply:
in that case why not make `_blas` a kwarg instead